### PR TITLE
Some changes

### PR DESCRIPTION
**Pull request by Joao-Victor-Bonfim**
**Please, review with care.**
 First of all, I would like to talk about separating the mod pack into modules.

 Having all rule sets in one file for each type makes searching through them and contributing a giant hassle. It also makes easier for errors to go unnoticed, like how Muton Guards unlocked alien medics, even though the string was an asset reuse. Therefore I added a markdown file (CONTRIBUTING.md) containing what boils down to a contribution guideline framework, feel free to modify it as you please.

 Beyond starting to modularize the mod pack. I simplified the medic research tree. Basically, now alien medics depends on whether an alien medic was interrogated or not, instead of depending on all alien medics. The medi-kit prerequisite became "alien medic interrogated" instead. the dependency graph should allow for more alien species medics to be added with more ease, as now you don't need to remember to add to their unlock list all research that is shared among all of them. I recommend doing the same for all other roles (engineer, soldier, leader, commander, guard and the like), which would allow for further research tree optimization, like: making so "alien is interrogated" is dependant on any alien role being interrogated, making so alien origins becomes less of a research mess and, if we follow the same logic, we could change the subterranean alien invasion to point to the alien underground species instead.

 I also rearranged the jarhead entry, because it was right on the top for some reason.

 Feel free to cherry pick what changes you like.

EDIT: also, I forgot to mention that I did not translate for all languages the medi-kit prerequisite change. I only did for the Latin languages and English.

### CODE REVIEW

**Comment by Joao-Victor-Bonfim** [comment]
Also, don't close the request right away, if you dislike its contents, we can discuss changes before merging, and you can wait to see them implemented first.

**Comment by SolariusScorch** [comment]
Hi Joao,

OK, please explain bout the modularization:
1) What do you mean by modularization? I see you changed some paths, but I don't really get the whole idea.
2) I completely fail to understand what's wrong with the current state of XCF. It has 7 MB in 47 files, which is pretty comfortable. For comparison, Piratez has 9 MB in 17 files (and more than half of them are very small, under 100 KB). I can't see how any improvement is necessary here.

Regarding the Medic change, I like it. The current state is a relic of very old architecture and simplifying it is a good idea.

The current Jarhead entry is fine, since it's about rank, not species. But it's totally understandable, sometimes I get confused too... :) I'm not 100% happy with how it works, but this system has to cover everything, from EXALT operators to Tentaculats.

Thanks for the input and please respond :)

Solar

**Comment by Joao-Victor-Bonfim** [comment]
> 
> 
> Hi Joao,
> 
> OK, please explain bout the modularization:
> 
>     1. What do you mean by modularization? I see you changed some paths, but I don't really get the whole idea.
> 
>     2. I completely fail to understand what's wrong with the current state of XCF. It has 7 MB in 47 files, which is pretty comfortable. For comparison, Piratez has 9 MB in 17 files (and more than half of them are very small, under 100 KB). I can't see how any improvement is necessary here.
> 
> 
> Regarding the Medic change, I like it. The current state is a relic of very old architecture and simplifying it is a good idea.
> 
> The current Jarhead entry is fine, since it's about rank, not species. But it's totally understandable, sometimes I get confused too... :) I'm not 100% happy with how it works, but this system has to cover everything, from EXALT operators to Tentaculats.
> 
> Thanks for the input and please respond :)
> 
> Solar

Well...
Modularization is the act of separating pieces of code into organized independent modules.
[Wikipedia article](https://en.wikipedia.org/wiki/Modular_programming)

This is the main reason for why I am proposing it.
> This makes modular designed systems, if built correctly, far more reusable than a traditional monolithic design, since all (or many) of these modules may then be reused (without change) in other projects. This also facilitates the "breaking down" of projects into several smaller projects. Theoretically, a modularized software project will be more easily assembled by large teams, since no team members are creating the whole system, or even need to know about the system as a whole. They can focus just on the assigned smaller task (this, it is claimed, counters the key assumption of The Mythical Man Month, making it actually possible to add more developers to a late software project without making it later still). 

And because problems can be pin pointed to fewer files, what makes managing changes less stressful.
You can also remove or refactor entire segments of code without forgetting anything, because it is contained within a set of folders/files.
It allows for parts of the mod to be isolated for future use on other mods without stress.

In sum: you can make your mod easier to maintain and expand by structuring it in a certain way.

File: Ruleset/research_XCOMFILES.rul [hunk 1]
```diff
@@ -5998,13 +5996,14 @@
     destroyItem: true
     unlocks:
       - STR_ALIEN_POWER_SYSTEMS_DEP_1
-      - STR_ALIEN_MEDICS
+      - STR_ALIEN_MEDIC_INTERROGATED
```
Comment (Joao-Victor-Bonfim) [inline Ruleset/research_XCOMFILES.rul:5999]: Also, delet this!
Ugh!

Comment (SolariusScorch) [reply Ruleset/research_XCOMFILES.rul:5999]: I still have no idea what modularization is and why it would be good. Please give me a real answer or stop wasting my time. I am trying to get engaged, I appreciate your effort, but please also respect mine.

From the article excerpt:
> This makes modular designed systems, if built correctly, far more reusable than a traditional monolithic design
I have no idea what it means.
> This also facilitates the "breaking down" of projects into several smaller projects.
I am not interested in that, XCF is one project, managed by one person. Any extra content goes through me anyway.
> And because problems can be pin pointed to fewer files, what makes managing changes less stressful.
I am completely unstressed as I am now.
> You can also remove or refactor entire segments of code without forgetting anything, because it is contained within a set of folders/files.
No idea what you're referring to.
> It allows for parts of the mod to be isolated for future use on other mods without stress.
No, it will be mess no other modder will want to touch.
Look, I know many mothers, some of them capable programmers, and nobody uses these methods.

Also:
STR_ALIEN_MEDICS - why would it be deleted?

Comment (Joao-Victor-Bonfim) [reply Ruleset/research_XCOMFILES.rul:5999]: > Also:
> STR_ALIEN_MEDICS - why would it be deleted?

Because it is under muton medic, which are muton guards.
I am just saying to also delete my addition.

**Comment by Joao-Victor-Bonfim** [comment]
> I still have no idea what modularization is and why it would be good. Please give me a real answer or stop wasting my time. I am trying to get engaged, I appreciate your effort, but please also respect mine.

It is because I have trouble explaining it, sorry.
;w;

> From the article excerpt:
> 
> > This makes modular designed systems, if built correctly, far more reusable than a traditional monolithic design
> 
> I have no idea what it means.

It means that, if you spread out the content into different files and different modules, it becomes far more useable than a system that is designed for everything to be in one definition and/or file.
 
> > This also facilitates the "breaking down" of projects into several smaller projects.
> 
> I am not interested in that, XCF is one project, managed by one person. Any extra content goes through me anyway.

But what if you want to use content from XCF in the future?
You would need to go through everything, remember what each piece of code does, then see in every other file if there are dependencies and so on.
When, instead, if you separate the code into boxes that are labelled, you can just go through the labels and take the entire folder without worrying if something that is unrelated is in there.

> > And because problems can be pin pointed to fewer files, what makes managing changes less stressful.
> 
> I am completely unstressed as I am now.

But bugs are still hard to find when looking through the code, as you need to go through thousands of lines of code to find where the problem is located.

> > You can also remove or refactor entire segments of code without forgetting anything, because it is contained within a set of folders/files.
> 
> No idea what you're referring to.
Let's say you just reworked the part of the research tree for the deep ones, how are you going to make sure you didn't miss anything if all the deep one research is all over the place on a single file?

if it was separated into a single folder called "Deep Ones" you would be sure that nothing unrelated to the deep ones would be inside that folder and that all content related to the deep ones would be contained within that folder.

> > It allows for parts of the mod to be isolated for future use on other mods without stress.
> 
> No, it will be mess no other modder will want to touch.
> Look, I know many mothers, some of them capable programmers, and nobody uses these methods.

It is a mess that no other modder would like to touch, because you aren't sure if you are forgetting anything if you try to pull a part of this mod pack out to insert into other mods. Without talking about the fact that modularization is the standard for programming, instead of the philosophy that is implemented in the XCF, that is a monolithic one where everything is contained in very few files.

**Comment by Joao-Victor-Bonfim** [comment]
And for the definition of modularization, in layman terms, is separating the code into organized little boxes of code that can communicate with each other, but aren't tied together. This separation allows for the programmer to treat the code as how it might be expected to behave (since it is separated into boxes, different parts that represent different things, like the GUI of a game and the AI aspect of it, can be worked on individually instead of being worked on all together). So instead of having every single research separated into a box called "research" we have research separated into different boxes that can speak with one another (dependencies), but can also be worked on individually, allowing for a better focus on it, which in turn increases productivity, reduces the chance for errors to appear and makes so that the addition or removal of new code can be treated individually without having to know or think about the inner workings of the other parts of code. Because you know that if you create a dependency with that part of code, it will say on it's usage what it does, and you can expect it to do just that.

We basically treat each part of the mod pack as a self contained opaque box.

File: Ruleset/research_XCOMFILES.rul (hunk 1 above)
Comment (SolariusScorch) [reply Ruleset/research_XCOMFILES.rul:5999]: Ah, I see now!

**Comment by SolariusScorch** [comment]
Frankly I still don't get your point, because you don't say anything concrete... But from your changes I understand that you want to divide the soldiers section into different parts, respective for the various "races".
Still, I find it impractical and I don't object to cutting up a section into several parts, as it's harder to work with - at least for me.
I did however fix the bugs and renamed that research.
Alien Medics stay as they are, because all other ranks work the same way, so why would medics be different? That's a good example of inconsistency which makes work harder...
For now closing this commit, but I can reopen if you wish.
